src/middleware/TelemetryHandler.ts

```diff
@@ -66,21 +66,35 @@ export class TelemetryHandler implements Middleware {
 	 */
 	public async execute(context: Context): Promise<void> {
 		try {
-			let clientRequestId: string = getRequestHeader(context.request, context.options, TelemetryHandler.CLIENT_REQUEST_ID_HEADER);
-			if (clientRequestId === null) {
-				clientRequestId = generateUUID();
-				setRequestHeader(context.request, context.options, TelemetryHandler.CLIENT_REQUEST_ID_HEADER, clientRequestId);
-			}
-			let sdkVersionValue: string = `${TelemetryHandler.PRODUCT_NAME}/${PACKAGE_VERSION}`;
-			let options: TelemetryHandlerOptions;
-			if (context.middlewareControl instanceof MiddlewareControl) {
-				options = context.middlewareControl.getMiddlewareOptions(TelemetryHandlerOptions) as TelemetryHandlerOptions;
-			}
-			if (typeof options !== "undefined") {
-				const featureUsage: string = options.getFeatureUsage();
-				sdkVersionValue += ` (${TelemetryHandler.FEATURE_USAGE_STRING}=${featureUsage})`;
+			if (typeof context.request === "string") {
+				if (isGraphURL(context.request)) {
+					// Add telemetry only if the request url is a Graph URL.
+					// Errors are reported as in issue #265 if headers are present when redirecting to a non Graph URL
+					let clientRequestId: string = getRequestHeader(context.request, context.options, TelemetryHandler.CLIENT_REQUEST_ID_HEADER);
+					if (clientRequestId === null) {
+						clientRequestId = generateUUID();
+						setRequestHeader(context.request, context.options, TelemetryHandler.CLIENT_REQUEST_ID_HEADER, clientRequestId);
+					}
+					let sdkVersionValue: string = `${TelemetryHandler.PRODUCT_NAME}/${PACKAGE_VERSION}`;
+					let options: TelemetryHandlerOptions;
+					if (context.middlewareControl instanceof MiddlewareControl) {
+						options = context.middlewareControl.getMiddlewareOptions(TelemetryHandlerOptions) as TelemetryHandlerOptions;
+					}
+					if (typeof options !== "undefined") {
+						const featureUsage: string = options.getFeatureUsage();
+						sdkVersionValue += ` (${TelemetryHandler.FEATURE_USAGE_STRING}=${featureUsage})`;
+					}
+					appendRequestHeader(context.request, context.options, TelemetryHandler.SDK_VERSION_HEADER, sdkVersionValue);
+				} else {
+					// Remove telemetry headers if present during redirection.
+					if (context.options.headers[TelemetryHandler.CLIENT_REQUEST_ID_HEADER]) {
+						delete context.options.headers[TelemetryHandler.CLIENT_REQUEST_ID_HEADER];
+					}
+					if (context.options.headers[TelemetryHandler.SDK_VERSION_HEADER]) {
+						delete context.options.headers[TelemetryHandler.SDK_VERSION_HEADER];
+					}
```

Comment:
I think you can call delete on the keys even if they don't exist. See [this example](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBDIKMjJHooWg-QTwLwAb0UJRKnAYyHyglQlE4zHZHIYzEgAH58px2IhpD9FABfADcwQQnD51UZzNA+ToWq8AF5QGyOYNuZBdQAiYhISAAOXNlt4wpKootoAAjAAGRUqxRqpnuALQeDMAAUGyZnAAlCruDBIApNb9mQBtS1uy0AXRVAfgQZD4cjzNjQA)
```suggestion
					delete context.options.headers[TelemetryHandler.CLIENT_REQUEST_ID_HEADER];
					delete context.options.headers[TelemetryHandler.SDK_VERSION_HEADER];
```